README.md
[![Build Status](https://travis-ci.org/advanced-rest-client/xml-viewer.svg?branch=stage)](https://travis-ci.org/advanced-rest-client/xml-viewer)

## UiElements.XmlViewer component
Tag: `<xml-viewer>`

### Installation
Using bower:
```
bower install --save advanced-rest-client/xml-viewer
```

`<xml-viewer>` An XML payload viewer for the XML response

### Example
```
<xml-viewer xml="&lt;tag&gt;&lt;/tag&gt;"></xml-viewer>
```

**Note** This element uses web workers with dependencies. It expect to find
workers files in current directory in the `workers` folder.
Your build process has to ensure that this files are avaiable.

## Content actions

You can add action items in the actions bar by adding elements as a children
of this element with slot set to `content-action`.

### Example
```
<xml-viewer>
 <paper-icon-button title="Additional action" icon="arc:cached" slot="content-action"></paper-icon-button>
 <paper-icon-button title="Clear the code" icon="arc:clear" slot="content-action"></paper-icon-button>
</xml-viewer>
```

## Changes in version 2

- The element doesn't mixin text search behavior. This service is deprecated.
- It uses worker files instead of compiled worker data in elements body

### Styling

`<xml-viewer>` provides the following custom properties and mixins for styling:

Custom property | Description | Default
----------------|-------------|----------
`--xml-viewer` | Mixin applied to the element | `{}`
`--xml-viewer-comment-color` | Color of the comment section. | `#236E25`
`--xml-viewer-punctuation-color` | Color of the punctuation signs | `black`
`--xml-viewer-tag-name-color` | Color of the XML tag name | `#881280`
`--xml-viewer-attribute-name-color` | Color of the XML attribute. | `#994500`
`--xml-viewer-attribute-value-color` | Color of the attribute's value. | `#1A1AA6`
`--xml-viewer-cdata-color` | CDATA section color. | `#48A`
`--xml-viewer-document-declaration-color` | XML document declaration (header) color. | `#999`
`--xml-viewer-constant-color` | Constant (boolean, null, number) color. | `#283593`

## API
### Component properties (attributes)

#### xml
- Type: `Object`
XML data to parse and display

#### isError
- Type: `boolean`
- Default: `false`
- Read only property
True if error ocurred when parsing data

#### working
- Type: `boolean`
- Default: `false`
- Read only property
True when XML is parsing

#### showOutput
- Type: `boolean`
- Default: `false`
- Read only property
True when output should be shown.

#### errorMessage
- Type: `string`
- Read only property
An error message to display.


### Component methods

#### disconnectedCallback
- Return type: `undefined`

#### reset
- Return type: `undefined`
Resets the state of the component.
#### render
- Return type: `undefined`
Parses and renders XML data.



CHANGELOG.md
<a name="2.0.3"></a>
## [2.0.3](https://github.com/advanced-rest-client/xml-viewer/compare/2.0.2...2.0.3) (2018-05-22)


### Update

* Bumping patch version ([93acf4f05b701341beea63a6e673be60b752aa02](https://github.com/advanced-rest-client/xml-viewer/commit/93acf4f05b701341beea63a6e673be60b752aa02))



<a name="2.0.3"></a>
## [2.0.3](https://github.com/advanced-rest-client/xml-viewer/compare/2.0.2...2.0.3) (2018-05-18)




<a name="2.0.2"></a>
## [2.0.2](https://github.com/advanced-rest-client/xml-viewer/compare/0.1.4...2.0.2) (2018-03-11)




<a name="0.1.5"></a>
## [0.1.5](https://github.com/advanced-rest-client/xml-viewer/compare/0.1.4...v0.1.5) (2017-04-28)




<a name="0.1.4"></a>
## [0.1.4](https://github.com/advanced-rest-client/xml-viewer/compare/0.1.2...v0.1.4) (2017-04-28)


### Fix

* Added promise polyfill in test for IE ([b145fa93369946a4a71bf249353b4f6c85ac63ce](https://github.com/advanced-rest-client/xml-viewer/commit/b145fa93369946a4a71bf249353b4f6c85ac63ce))

### Update

* Removed linter from Travis test due to pass the test after inlining web workers code into the content of the element ([09a9c3509011f8709350e8f56e59a80c16e35909](https://github.com/advanced-rest-client/xml-viewer/commit/09a9c3509011f8709350e8f56e59a80c16e35909))
* Replaced external web worker files with worker content included into the element body ([89c6f059813e40ef6d2f38f69277a3bd3879c0b8](https://github.com/advanced-rest-client/xml-viewer/commit/89c6f059813e40ef6d2f38f69277a3bd3879c0b8))



<a name="0.1.3"></a>
## [0.1.3](https://github.com/advanced-rest-client/xml-viewer/compare/0.1.2...v0.1.3) (2017-03-13)




<a name="0.1.2"></a>
## [0.1.2](https://github.com/advanced-rest-client/xml-viewer/compare/0.1.1...v0.1.2) (2017-03-13)


### Docs

* Updated documentation for cvontent actions ([8e322629a66028617faec90adad1e3091dfc9016](https://github.com/advanced-rest-client/xml-viewer/commit/8e322629a66028617faec90adad1e3091dfc9016))

### Fix

* Added fetch polyfil to test for Safari - the new IE ([c19532d52d907f954332ddb39d8b1ac7a7bbaddb](https://github.com/advanced-rest-client/xml-viewer/commit/c19532d52d907f954332ddb39d8b1ac7a7bbaddb))
* Fixed importLocation variable initialization to work in all cases ([5ca7bd2ce9155ff68a8360dbca39dd254bff8179](https://github.com/advanced-rest-client/xml-viewer/commit/5ca7bd2ce9155ff68a8360dbca39dd254bff8179))

### New

* Added importLocation element's property ([376ad6124dee0cfb1279c21fd16bac270fcbb305](https://github.com/advanced-rest-client/xml-viewer/commit/376ad6124dee0cfb1279c21fd16bac270fcbb305))
* Added placeholder for content actions ([665696a630c4bf0b63eccc5a27444d0acc2186c2](https://github.com/advanced-rest-client/xml-viewer/commit/665696a630c4bf0b63eccc5a27444d0acc2186c2))

### Update

* Added missing dependency ([eba74751c71ee723ef22dd3e6bac61f1d00e964f](https://github.com/advanced-rest-client/xml-viewer/commit/eba74751c71ee723ef22dd3e6bac61f1d00e964f))
* Updated demo page for content actions API ([8a47c3c22856a9b45f89040dc56129cb2b301e72](https://github.com/advanced-rest-client/xml-viewer/commit/8a47c3c22856a9b45f89040dc56129cb2b301e72))
* Updated test cases for content action ([7e50646e0aa604723902e1b60b9113cb9b8c372c](https://github.com/advanced-rest-client/xml-viewer/commit/7e50646e0aa604723902e1b60b9113cb9b8c372c))
* Updated Travis configuration for Sauce ([38b03426e7d58a69513fc303ea56cbafe7f09814](https://github.com/advanced-rest-client/xml-viewer/commit/38b03426e7d58a69513fc303ea56cbafe7f09814))



<a name="0.1.1"></a>
## 0.1.1 (2017-01-03)




# [2.1.0](https://github.com/advanced-rest-client/xml-viewer/compare/2.0.2...2.1.0) (2019-02-14)


### Docs

* Updating polymer docs and typpings ([ec29d5586db7bcbfb17c57a573be8667df54f11b](https://github.com/advanced-rest-client/xml-viewer/commit/ec29d5586db7bcbfb17c57a573be8667df54f11b))

### Fix

* Adding missing dependencies to travis ([179097921e27e3a772028279f3aa7f673160e130](https://github.com/advanced-rest-client/xml-viewer/commit/179097921e27e3a772028279f3aa7f673160e130))
* Fixing test command ([3fa52e4f76c4cea62eeecac0d62f870d6b908568](https://github.com/advanced-rest-client/xml-viewer/commit/3fa52e4f76c4cea62eeecac0d62f870d6b908568))

### New

* Adding deprecation message. ([b440932affabf1166554258c42d1440e7abfb4e9](https://github.com/advanced-rest-client/xml-viewer/commit/b440932affabf1166554258c42d1440e7abfb4e9))

### Update

* Bumping patch version ([93acf4f05b701341beea63a6e673be60b752aa02](https://github.com/advanced-rest-client/xml-viewer/commit/93acf4f05b701341beea63a6e673be60b752aa02))
* Removing old CI files ([ea57a4f5f04363a9624b9c39fc9686a0ba3b93a2](https://github.com/advanced-rest-client/xml-viewer/commit/ea57a4f5f04363a9624b9c39fc9686a0ba3b93a2))
* The XML is being rendered despite the error. ([f7f074d9aadaf1d37968cf0bcd6290a190e9d326](https://github.com/advanced-rest-client/xml-viewer/commit/f7f074d9aadaf1d37968cf0bcd6290a190e9d326))
* Updating reset behavior ([a4785f90076e2d2a8e58ec5c9f9aa15558995330](https://github.com/advanced-rest-client/xml-viewer/commit/a4785f90076e2d2a8e58ec5c9f9aa15558995330))
* Updating tests, adding coverage ([d9c7fbda2ff8437fb0698e376eac9b4c6b4f47c7](https://github.com/advanced-rest-client/xml-viewer/commit/d9c7fbda2ff8437fb0698e376eac9b4c6b4f47c7))



## [2.1.1](https://github.com/advanced-rest-client/xml-viewer/compare/2.1.0...2.1.1) (2019-02-20)




# [3.0.0](https://github.com/advanced-rest-client/xml-viewer/compare/2.1.0...3.0.0) (2019-07-29)





test/index.d.ts
/**
 * DO NOT EDIT
 *
 * This file was automatically generated by
 *   https://github.com/Polymer/gen-typescript-declarations
 *
 * To modify these typings, edit the source file(s):
 *   test/index.html
 */

/// <reference path="../../webcomponentsjs/webcomponents-lite.d.ts" />
/// <reference path="../../web-component-tester/browser.d.ts" />



test/basic-test.d.ts
/**
 * DO NOT EDIT
 *
 * This file was automatically generated by
 *   https://github.com/Polymer/gen-typescript-declarations
 *
 * To modify these typings, edit the source file(s):
 *   test/basic-test.html
 */

/// <reference path="../../webcomponentsjs/webcomponents-lite.d.ts" />
/// <reference path="../../web-component-tester/browser.d.ts" />
/// <reference path="../xml-viewer.d.ts" />
/// <reference path="../../paper-icon-button/paper-icon-button.d.ts" />



test/xml-viewer.test.js
import { fixture, assert, nextFrame } from '@open-wc/testing';
import sinon from 'sinon/pkg/sinon-esm.js';
import '@polymer/paper-icon-button/paper-icon-button.js';
import '../xml-viewer.js';


describe('<xml-viewer>', () => {
  async function basicFixture() {
    return (await fixture(`<xml-viewer></xml-viewer>`));
  }

  async function contentActionFixture() {
    return (await fixture(`<xml-viewer>
      <paper-icon-button slot="content-action"></paper-icon-button>
    </xml-viewer>`));
  }

  async function getXml(file) {
    return new Promise((resolve, reject) => {
      const xhr = new XMLHttpRequest();
      xhr.open('GET', file, true);
      xhr.send();
      xhr.onload = (e) => {
        resolve(e.target.response);
      };
      xhr.onerror = () => {
        reject(new Error('Unable to get test xml data'));
      };
    });
  }

  describe('basic', () => {
    let xml;
    let element;

    before(async () => {
      xml = await getXml('/base/demo/valid.xml');
    });

    beforeEach(async () => {
      element = await basicFixture();
      element.xml = xml;
    });

    it('Should parse HTML', function() {
      const output = element.shadowRoot.querySelector('output');
      const html = output.innerHTML;
      assert.isString(html);
    });
  });

  describe('content actions', () => {
    let element;
    beforeEach(async () => {
      element = await contentActionFixture();
    });

    it('Should have distributed nodes', () => {
      const nodes = element.shadowRoot.querySelector('slot').assignedNodes();
      assert.equal(nodes.length, 1);
    });
  });

  describe('_changed()', () => {
    let xml;
    let element;
    before(async () => {
      xml = await getXml('/base/demo/valid.xml');
    });

    beforeEach(async () => {
      element = await basicFixture();
    });

    it('Function is called when "xml" property chnage', () => {
      const spy = sinon.spy(element, '_changed');
      element.xml = xml;
      assert.isTrue(spy.called);
    });

    it('Calls _renderXml() with argument', () => {
      const spy = sinon.spy(element, '_renderXml');
      element.xml = xml;
      assert.equal(spy.args[0][0], xml);
    });

    it('Calls reset() when no XML', () => {
      const spy = sinon.spy(element, 'reset');
      element._changed();
      assert.isTrue(spy.called);
    });
  });

  describe('reset()', () => {
    let element;
    beforeEach(async () => {
      element = await basicFixture();
    });

    it('Sets working property', () => {
      element._working = true;
      element.reset();
      assert.isFalse(element._working);
    });

    it('Sets isError property', () => {
      element._isError = true;
      element.reset();
      assert.isFalse(element._isError);
    });

    it('Clears errorMessage', () => {
      element._errorMessage = 'test';
      element.reset();
      assert.equal(element._errorMessage, null);
    });

    it('Clears the output', () => {
      element._output.innerText = 'test';
      element.reset();
      assert.equal(element._output.innerText, '');
    });
  });

  describe('_renderXml()', () => {
    let xml;
    let element;
    before(async () => {
      xml = await getXml('/base/demo/valid.xml');
    });

    beforeEach(async () => {
      element = await basicFixture();
    });

    it('Calls _processData()', () => {
      const spy = sinon.spy(element, '_processData');
      element._renderXml(xml);
      assert.isTrue(spy.called);
    });
  });

  describe('Errored XML', () => {
    let xml;
    let element;
    before(async () => {
      xml = await getXml('/base/demo/invalid.xml');
    });

    beforeEach(async () => {
      element = await basicFixture();
      element.xml = xml;
      await nextFrame();
    });

    it('Renders error message', () => {
      const error = element.shadowRoot.querySelector('error-message');
      assert.ok(error);
    });

    it('Sets _isError', () => {
      assert.isTrue(element._isError);
    });

    it('Sets _working', () => {
      assert.isFalse(element._working);
    });

    it('Sets _errorMessage', () => {
      assert.typeOf(element._errorMessage, 'string');
    });

    it('Sets output', () => {
      const txt = element._output.innerText.trim();
      assert.equal(txt.indexOf('<feed xml:base'), 0);
    });
  });

  describe('_handleDisplayClick()', () => {
    let element;
    beforeEach(async () => {
      element = await basicFixture();
      element._output.innerHTML = `
      <div class="top">
        <div class="expansion-not-set">
          <div class="parent-collapse-marker" colapse-marker="true">
            <div class="child-collapse-marker"></div>
          </div>
        </div>
        <div class="expansion-set-true" data-expanded="true">
          <div class="parent-collapse-marker-set-true" colapse-marker="true">
            <div class="child-collapse-marker-set-true"></div>
          </div>
        </div>
        <div class="expansion-set-false" data-expanded="false">
          <div class="parent-collapse-marker-set-false" colapse-marker="true">
            <div class="child-collapse-marker-set-false"></div>
          </div>
        </div>
      </div>
      `;
    });

    it('Does nothing when no target', () => {
      element._handleDisplayClick({});
      // no error
    });

    it('Toggles unset from child', () => {
      const node = element._output.querySelector('.child-collapse-marker');
      node.click();
      const target = element._output.querySelector('.expansion-not-set');
      assert.equal(target.dataset.expanded, 'false');
    });

    it('Toggles unset from parent', () => {
      const node = element._output.querySelector('.parent-collapse-marker');
      node.click();
      const target = element._output.querySelector('.expansion-not-set');
      assert.equal(target.dataset.expanded, 'false');
    });

    it('Toggles set to "true" from child', () => {
      const node = element._output.querySelector('.child-collapse-marker-set-true');
      node.click();
      const target = element._output.querySelector('.expansion-set-true');
      assert.equal(target.dataset.expanded, 'false');
    });

    it('Toggles set to "true" from parent', () => {
      const node = element._output.querySelector('.parent-collapse-marker-set-true');
      node.click();
      const target = element._output.querySelector('.expansion-set-true');
      assert.equal(target.dataset.expanded, 'false');
    });

    it('Toggles set to "false" from child', () => {
      const node = element._output.querySelector('.child-collapse-marker-set-false');
      node.click();
      const target = element._output.querySelector('.expansion-set-false');
      assert.equal(target.dataset.expanded, 'true');
    });

    it('Toggles set to "false" from parent', () => {
      const node = element._output.querySelector('.parent-collapse-marker-set-false');
      node.click();
      const target = element._output.querySelector('.expansion-set-false');
      assert.equal(target.dataset.expanded, 'true');
    });
  });
});


karma.sl.config.js
/* eslint-disable import/no-extraneous-dependencies */
const merge = require('webpack-merge');
const slSettings = require('@advanced-rest-client/testing-karma-sl/sl-settings.js');
const createBaseConfig = require('./karma.conf.js');

module.exports = (config) => {
  const cnf = slSettings();
  cnf.sauceLabs.testName = 'xml-viewer';
  cnf.browsers = [
    'SL_Chrome',
    'SL_Chrome-1',
    'SL_Firefox',
    'SL_Firefox-1',
    'SL_Safari',
    'SL_EDGE'
  ];
  if (process.env.TRAVIS) {
    const buildLabel = 'TRAVIS #' + process.env.TRAVIS_BUILD_NUMBER + ' (' + process.env.TRAVIS_BUILD_ID + ')';

    cnf.browserStack = {
      build: buildLabel,
      tunnelIdentifier: process.env.TRAVIS_JOB_NUMBER
    };

    cnf.sauceLabs.build = buildLabel;
    cnf.sauceLabs.tunnelIdentifier = process.env.TRAVIS_JOB_NUMBER;
  }

  config.set(merge(createBaseConfig(config), cnf));

  return config;
};


xml-viewer.js
/**
@license
Copyright 2016 The Advanced REST client authors <arc@mulesoft.com>
Licensed under the Apache License, Version 2.0 (the "License"); you may not
use this file except in compliance with the License. You may obtain a copy of
the License at
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
License for the specific language governing permissions and limitations under
the License.
*/
import { LitElement, html, css } from 'lit-element';
import '@polymer/paper-spinner/paper-spinner.js';
import '@polymer/iron-icon/iron-icon.js';
import '@advanced-rest-client/arc-icons/arc-icons.js';
import '@advanced-rest-client/error-message/error-message.js';

const SafeHtmlUtils = {
  AMP_RE: new RegExp(/&/g),
  GT_RE: new RegExp(/>/g),
  LT_RE: new RegExp(/</g),
  SQUOT_RE: new RegExp(/'/g),
  QUOT_RE: new RegExp(/"/g),
  htmlEscape: function(s) {
    if (s.indexOf('&') !== -1) {
      s = s.replace(SafeHtmlUtils.AMP_RE, '&amp;');
    }
    if (s.indexOf('<') !== -1) {
      s = s.replace(SafeHtmlUtils.LT_RE, '&lt;');
    }
    if (s.indexOf('>') !== -1) {
      s = s.replace(SafeHtmlUtils.GT_RE, '&gt;');
    }
    if (s.indexOf('"') !== -1) {
      s = s.replace(SafeHtmlUtils.QUOT_RE, '&quot;');
    }
    if (s.indexOf('\'') !== -1) {
      s = s.replace(SafeHtmlUtils.SQUOT_RE, '&#39;');
    }
    return s;
  }
};
/**
 * `<xml-viewer>` An XML payload viewer for the XML response
 *
 * ### Example
 * ```
 * <xml-viewer xml="&lt;tag&gt;&lt;/tag&gt;"></xml-viewer>
 * ```
 *
 * **Note** This element uses web workers with dependencies. It expect to find
 * workers files in current directory in the `workers` folder.
 * Your build process has to ensure that this files are avaiable.
 *
 * ## Content actions
 *
 * You can add action items in the actions bar by adding elements as a children
 * of this element with slot set to `content-action`.
 *
 * ### Example
 * ```
 * <xml-viewer>
 *  <paper-icon-button title="Additional action" icon="arc:cached" slot="content-action"></paper-icon-button>
 *  <paper-icon-button title="Clear the code" icon="arc:clear" slot="content-action"></paper-icon-button>
 * </xml-viewer>
 * ```
 *
 * ## Changes in version 2
 *
 * - The element doesn't mixin text search behavior. This service is deprecated.
 * - It uses worker files instead of compiled worker data in elements body
 *
 * ### Styling
 *
 * `<xml-viewer>` provides the following custom properties and mixins for styling:
 *
 * Custom property | Description | Default
 * ----------------|-------------|----------
 * `--xml-viewer` | Mixin applied to the element | `{}`
 * `--xml-viewer-comment-color` | Color of the comment section. | `#236E25`
 * `--xml-viewer-punctuation-color` | Color of the punctuation signs | `black`
 * `--xml-viewer-tag-name-color` | Color of the XML tag name | `#881280`
 * `--xml-viewer-attribute-name-color` | Color of the XML attribute. | `#994500`
 * `--xml-viewer-attribute-value-color` | Color of the attribute's value. | `#1A1AA6`
 * `--xml-viewer-cdata-color` | CDATA section color. | `#48A`
 * `--xml-viewer-document-declaration-color` | XML document declaration (header) color. | `#999`
 * `--xml-viewer-constant-color` | Constant (boolean, null, number) color. | `#283593`
 *
 * @customElement
 * @polymer
 * @memberof UiElements
 * @demo demo/index.html
 */
export class XmlViewer extends LitElement {
  static get styles() {
    return css`:host {
      display: block;
      color: black;
      cursor: text;
      user-select: text;
      overflow: auto;
      word-wrap: break-word;
    }

    [hidden] {
      display: none !important;
    }

    .material-icons {
      font-family: 'Material Icons';
      font-weight: normal;
      font-style: normal;
      font-size: 24px;
      /* Preferred icon size */
      display: inline-block;
      line-height: 1;
      text-transform: none;
      letter-spacing: normal;
      word-wrap: normal;
      white-space: nowrap;
      direction: ltr;
      -webkit-font-smoothing: antialiased;
      text-rendering: optimizeLegibility;
      -moz-osx-font-smoothing: grayscale;
    }

    .prettyPrint {
      font-family: monospace;
      font-size: 15px;
    }

    .arrowEmpty {}

    .node {
      margin: 1px 0px;
    }

    .opened {}

    .comment {
      color: var(--xml-viewer-comment-color, #236E25);
    }

    .punctuation {
      color: var(--xml-viewer-punctuation-color, black);
    }

    .tagname {
      color: var(--xml-viewer-tag-name-color, #881280);
    }

    .attname {
      color: var(--xml-viewer-attribute-name-color, #994500);
    }

    .attribute {
      color: var(--xml-viewer-attribute-value-color, #1A1AA6);
    }

    .cdata {
      color: var(--xml-viewer-cdata-color, #48A);
    }

    .cdata *[collapsible] {
      white-space: pre;
    }

    .arrowExpanded,
    .arrowEmpty {
      display: inline-block;
      width: 24px;
      height: 18px;
    }

    .processing {
      color: var(--xml-viewer-document-declaration-color, #999);
    }

    .inline,
    .inline > div {
      display: inline-block;
      text-indent: 0px;
    }

    .node.opened > arrowEmpty {
      text-indent: 0;
      font-size: 10px;
      letter-spacing: 0.1em;
      width: 21px;
      margin-left: 3px;
      margin-right: 3px;
    }

    .nodeMargin {
      margin-left: 44px;
    }

    .collapseIndicator {
      display: none;
      margin: 0px 1px;
      text-indent: 0px;
    }

    *[colapse-marker] {
      -webkit-user-select: none;
      cursor: pointer;
    }

    *[less] {
      display: inline-block;
    }

    *[more] {
      display: none;
    }

    *[data-expanded="false"] .arrowEmpty {
      display: none;
    }

    *[data-expanded="false"] *[less] {
      display: none;
    }

    *[data-expanded="false"] *[more] {
      display: inline-block;
    }

    *[data-expanded="false"] *[collapsible] {
      display: none !important;
    }

    *[data-expanded="false"] *[collapse-indicator] {
      display: inline-block !important;
    }

    .value.number,
    .value.null,
    .value.boolean {
      color: var(--xml-viewer-constant-color, #283593);
    }

    .arc-search-mark.selected {
      background-color: #ff9632;
    }

    .actions-panel {
      display: flex;
      flex-direction: row;
      align-items: center;
    }

    .actions-panel.hidden {
      display: none;
    }

    .spinner {
      position: relative;
    }

    paper-spinner {
      position: absolute;
    }

    .deprecation {
      color: #9E9E9E;
    }`;
  }

  render() {
    const { xml, _working, _isError, errorMessage, hideDeprecationMessage } = this;
    const showOutput = !_working && !!xml;
    return html`
    <p ?hidden="${hideDeprecationMessage}" class="deprecation">
      This view is retired and will be replaced around June 2019. <a target="_blank" href="https://restforchrome.blogspot.com/2019/02/deprecation-of-xml-viewer-component.html">Learn more.</a>
    </p>
    <div class="${this._computeActionsPanelClass(showOutput)}">
      <slot name="content-action"></slot>
    </div>
    <div class="spinner">
      <paper-spinner .active="${_working}"></paper-spinner>
    </div>
    <error-message ?hidden="${!_isError}">
      <p>There was an error parsing XML.</p>
      <p>${errorMessage}</p>
      <p>Rendering unprocessed data.</p>
    </error-message>
    <output ?hidden="${!showOutput}" @click="${this._handleDisplayClick}"></output>`;
  }

  static get properties() {
    return {
      /**
       * XML data to parse and display
       */
      xml: { type: Object },
      /**
       * True if error ocurred when parsing data
       */
      _isError: { type: Boolean },
      /**
       * True when XML is parsing
       */
      _working: { type: Boolean },
      /**
       * An error message to display.
       */
      _errorMessage: { type: String },
      /**
       * When set deprecation message won't be rendered.
       */
      hideDeprecationMessage: { type: Boolean },
      /**
       * Used in generating HTML output to prefix CSS classes for CSS scopes.
       */
      cssPrefix: { type: String }
    };
  }

  get _output() {
    return this.shadowRoot.querySelector('output');
  }

  get xml() {
    return this._xml;
  }

  set xml(value) {
    const old = this._xml;
    if (old === value) {
      return;
    }
    this._xml = value;
    this._changed(value);
  }

  get isError() {
    return this._isError;
  }

  get _isError() {
    return this.__isError;
  }

  set _isError(value) {
    const old = this.__isError;
    if (old === value) {
      return;
    }
    this.__isError = value;
    this.requestUpdate('_isError', old);
    this.dispatchEvent(new CustomEvent('iserror-changed', {
      detail: {
        value
      }
    }));
  }

  get working() {
    return this._working;
  }

  get _working() {
    return this.___working;
  }

  set _working(value) {
    const old = this.___working;
    if (old === value) {
      return;
    }
    this.___working = value;
    this.requestUpdate('_working', old);
    this.dispatchEvent(new CustomEvent('working-changed', {
      detail: {
        value
      }
    }));
  }

  constructor() {
    super();
    this.cssPrefix = 'xml-viewer style-scope ';
  }

  /**
   * Handler for the xml attribute change.
   *
   * @param {String} xml Changed value.
   */
  _changed(xml) {
    if (!xml) {
      this.reset();
      return;
    }
    this._renderXml(xml);
  }
  /**
   * Resets the state of the component.
   */
  reset() {
    this._working = false;
    this._isError = false;
    this._errorMessage = null;
    const out = this._output;
    if (out) {
      this._output.innerText = '';
    }
  }
  /**
   * Parses and renders XML data.
   *
   * @param {String} xml XML string to parse and render.
   */
  _renderXml(xml) {
    this.reset();
    this._working = true;
    try {
      const data = this._processData(xml);
      this._working = false;
      this._output.innerHTML = data;
    } catch (e) {
      this._parsingError(e, xml);
    }
  }
  /**
   * Handles parsing errors
   *
   * @param {Error} e
   * @param {String} xml Original XML string
   */
  _parsingError(e, xml) {
    this._isError = true;
    this._working = false;
    const message = e.message || 'Invalid XML value.';
    const err = message.replace('Uncaught Error: ', '');
    this._errorMessage = err;
    const out = this._output;
    if (out) {
      this._output.innerText = xml || '';
    }
  }
  /**
   * Handles clicks on the rendered items.
   * Provides support for expand / collapse functions.
   *
   * @param {Event} e
   */
  _handleDisplayClick(e) {
    if (!e.target) {
      return;
    }
    let target = e.target;
    if (!target.getAttribute('colapse-marker')) {
      target = target.parentNode;
      if (!target || !target.getAttribute('colapse-marker')) {
        return;
      }
    }
    target = target.parentNode;
    const expanded = target.dataset.expanded;
    if (!expanded || expanded === 'true') {
      target.dataset.expanded = 'false';
    } else {
      target.dataset.expanded = 'true';
    }
  }

  // Computes CSS class for the content actions pane.
  _computeActionsPanelClass(showOutput) {
    let clazz = 'actions-panel';
    if (!showOutput) {
      clazz += ' hidden';
    }
    return clazz;
  }

  _processData(xml) {
    const parser = new DOMParser();
    const result = parser.parseFromString(xml, 'application/xml');
    const err = result.querySelector('parsererror');
    if (err) {
      const message = err.innerText;
      throw new Error(message);
    }
    return this._getHTML(result);
  }

  _getHTML(doc) {
    const nodes = doc.childNodes;
    const nodesCnt = nodes.length;
    if (nodesCnt === 0) {
      return 'no xml';
    }
    let result = '<div class="' + this.cssPrefix + 'prettyPrint" data-expanded="true">';
    for (let i = 0; i < nodesCnt; i++) {
      result += this._parse(nodes.item(i));
    }
    result += '</div>';
    return result;
  }

  _parse(node) {
    const cssPrefix = this.cssPrefix;
    let parsed = '';
    const type = node.nodeType;
    switch (type) {
      case 1:
        // ELEMENT_NODE, value null
        parsed += this._parseElement(node);
        break;
      case 3:
        // TEXT_NODE, content of node
        {
          let value = node.nodeValue;
          value = SafeHtmlUtils.htmlEscape(value);
          if (value === '') {
            return '';
          }
          parsed += this._parseValue(value);
        }
        break;
      case 4:
        // CDATA_SECTION_NODE, content of node
        parsed += '<span colapse-marker="true" class="' + cssPrefix +
          '"><iron-icon more icon="arc:expand-more" class="' + cssPrefix + '"></iron-icon>' +
          '<iron-icon less icon="arc:expand-less" class="' + cssPrefix + '"></iron-icon></span>';
        parsed += '<span class="' + cssPrefix + 'cdata">&lt;![CDATA[</span>';
        parsed += '<div collapsible class="' + cssPrefix + '">';
        // parsed += this._urlify(SafeHtmlUtils.htmlEscape(node.nodeValue));
        parsed += SafeHtmlUtils.htmlEscape(node.nodeValue);
        parsed += '</div><span class="' + cssPrefix + 'cdata">]]&gt;</span>';
        break;
      case 7:
        // document declaration
        parsed += '<div class="' + cssPrefix + 'processing">&lt;?xml ' + node.nodeValue +
          ' ?&gt;</div>';
        break;
      case 8:
        // COMMENT_NODE, comment text
        parsed += '<div class="' + cssPrefix + 'comment">&lt;--';
        parsed += node.nodeValue;
        parsed += '--&gt</div>';
        break;
    }
    parsed = '<div class="' + cssPrefix + 'node">' + parsed + '</div>';
    return parsed;
  }

  _parseValue(value) {
    value = value.trim();
    let css = 'value';
    if (!isNaN(value)) {
      css += ' number';
    } else if (value === 'true' || value === 'false') {
      css += ' boolean';
    } else if (value === 'null') {
      css += ' null';
    }
    value = '<span class="' + this.cssPrefix + css + '">' + value + '</span>';
    return value;
  }

  _parseElement(node) {
    const cssPrefix = this.cssPrefix;
    const childrenCount = node.childNodes.length;
    let parsed = '';
    let showArrows = false;

    if (childrenCount > 1) {
      parsed += '<span colapse-marker="true" class="' + cssPrefix +
        '"><iron-icon more icon="arc:expand-more" class="' + cssPrefix + '"></iron-icon>' +
        '<iron-icon less icon="arc:expand-less" class="' + cssPrefix + '"></iron-icon></span>';
      showArrows = true;
    }
    parsed += '<span class="' + cssPrefix + 'punctuation">&lt;</span>';
    parsed += '<span class="' + cssPrefix + 'tagname">' + node.nodeName + '</span>';
    parsed += this._parseAttributes(node);
    if (childrenCount > 0) {
      const children = node.childNodes;
      parsed += '<span class="' + cssPrefix + 'punctuation">&gt;</span>';

      let showInline = false;
      if (childrenCount === 1 && children.item(0).nodeType === 3) {
        // simple: only one child - text - show response inline.
        showInline = true;
      }
      if (showInline) {
        parsed += '<div class="' + cssPrefix + 'inline">';
      } else {
        parsed += '<div collapse-indicator colapse-marker="true" class="' + cssPrefix +
          'collapseIndicator">...</div>';
        parsed += '<div collapsible class="' + cssPrefix + 'nodeMargin">';
      }
      for (let i = 0; i < childrenCount; i++) {
        parsed += this._parse(children.item(i));
      }

      parsed += '</div>';

      if (showArrows) {
        parsed += '<span arrowEmpty class="' + cssPrefix + 'arrowEmpty">&nbsp;</span>';
      }
      parsed += '<span class="' + cssPrefix + 'punctuation end">&lt;/</span>';
      parsed += '<span class="' + cssPrefix + 'tagname end">' + node.nodeName + '</span>';
      parsed += '<span class="' + cssPrefix + 'punctuation">&gt;</span>';
    } else {
      parsed += '<span class="' + cssPrefix + 'punctuation"> /&gt;</span>';
    }
    return parsed;
  }

  _parseAttributes(node) {
    let parsed = '';
    const attr = node.attributes;
    if (attr !== null && attr.length > 0) {
      for (let i = 0; i < attr.length; i++) {
        parsed += ' ' + this._getAttributesString(attr.item(i));
      }
    }
    return parsed;
  }

  _getAttributesString(attr) {
    const cssPrefix = this.cssPrefix;
    let data = '<span class="' + cssPrefix + 'attname">';
    let name = attr.name;
    name = SafeHtmlUtils.htmlEscape(name);
    data += name;
    data += '</span>';
    data += '<span class="' + cssPrefix + 'punctuation">=</span>';
    data += '<span class="' + cssPrefix + 'attribute">&quot;';
    let value = attr.value;
    value = SafeHtmlUtils.htmlEscape(value);
    data += value;
    data += '&quot;</span>';
    return data;
  }
}
/* eslint-disable wc/no-invalid-element-name */
window.customElements.define('xml-viewer', XmlViewer);
